### ISSUE TITLE
WARNING:tensorflow:Gradients do not exist for variables ['p_re_lu/alpha:0'] when minimizing the loss.

### ISSUE THREAD

**Issue by beginner401** [keras-team/keras#15716]
**System information**.
Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-74-generic x86_64)
TensorFlow 2.0.0 installed by anaconda
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
NVIDIA-SMI 460.80 Driver Version: 460.80 CUDA Version: 11.2

**Describe the problem**.
Sometimes when I'm using tf.keras.layers.PReLU() with a constant initializer, it occurs that the gradients do not exist as the warning infomation says'WARNING:tensorflow:Gradients do not exist for variables ['p_re_lu/alpha:0'] when minimizing the loss.'. However, I'm not sure why it occurs as somtimes it occurs but somtimes not. Further more, I have used many other layers for testing with various strange structure, but the warning only appear on PReLU, so I suggest it has something wrong with PReLU, but not my network structure. Or there must be something special on PReLU. If I am wrong, please correct me, thanks!

**Describe the current behavior**.
Sometimes when I'm using tf.keras.layers.PReLU() with a constant initializer, it occurs that the gradients do not exist as the warning infomation says'WARNING:tensorflow:Gradients do not exist for variables ['p_re_lu/alpha:0'] when minimizing the loss.'. 

**Describe the expected behavior**.
Gradients for tf.keras.layers.PReLU() should always exists.

**Standalone code to reproduce the issue**.
Here is a network structure in json that will produce the bug:
{
	"input_shape": [28, 28, 1], 
	"network": [
		{"name": "Squeeze", "params": {"tensor_space": 4, "dim": 3}}, 
		{"name": "Threshold", "params": {"threshold": 0.9880963343359307}}, 
		{"name": "PReLU", "params": {"init": 0.4469039064898337, "share": true}}, 
		{"name": "BiasAdd", "params": {"bias": 0.021902653299097977}}, 
		{"name": "Sqrt", "params": {}}, {"name": "Sqrt", "params": {}}, 
		{"name": "Sqrt", "params": {}}, {"name": "Ceil", "params": {}}, 
		{"name": "ReduceSum", "params": {"keep_dims": false, "dim": 2, "tensor_space": 4}}, 
		{"name": "Softmax", "params": {"dim": -1}}, 
		{"name": "GaussianNoise", "params": {"stddev": 0.0052523236865178015}}, 
		{"name": "Dense", "params": {"in_features": 28, "out_features": 10}}, 
		{"name": "Softmax"}
	]
}
If there is not a layer class in tf.keras.layers related to the above ops, I encapsulate the function using tf.keras.layers.Layer, and the PReLU will be interpreted as:
initializer = tf.keras.initializers.Constant(alpha)
if share:
    shared = [i + 1 for i in range(len(input_shape))]
else:
    if len(input_shape) == 1:
        shared = None
    else:
        shared = [i + 1 for i in range(len(input_shape) - 1)]
p_relu = tf.keras.layers.PReLU(alpha_initializer=initializer, shared_axes=shared)

The training process is implemented using standardized code for 'with tf.GradientTape() as tape:'.

**Source code / logs**.
WARNING:tensorflow:Gradients do not exist for variables ['p_re_lu/alpha:0'] when minimizing the loss.

**Comment by chunduriv** [comment]
@beginner401, In order to expedite the trouble-shooting process, please provide a standalone code snippet to reproduce the issue reported here. Thanks!

**Comment by beginner401** [comment]
@chunduriv It seems that I made a mistake, and the problem is triggered on ceil, but not PReLU. Ceil will 'steal' the gradients of layers before it. So, considering ceil function is non-differentiable, the gradients of layers before it seems meaningless. However, some functions like that will approximate a value, ReLU, MaxPool, e.g. So, I'm curious why ceil does not have a category designed like that, for example, a linear approximation, is it beause ceil is designed to be meaningless in nerual network? Or am I using the wrong pattern to define the Ceil layer? Thanks!
If you still need a code, the code below will trigger the same problem:
import numpy as np
import tensorflow as tf


class Ceil(tf.keras.layers.Layer):
    def __init__(self):
        super(Ceil, self).__init__()

    def build(self, input_shape):
        super(Ceil, self).build(input_shape)

    def call(self, x):
        return tf.math.ceil(x)


(x_train, y_train), (_, _) = tf.keras.datasets.mnist.load_data()
x_train = x_train.reshape((-1, 28, 28, 1))
materials = np.eye(10)
y = materials[y_train]
x = x_train
x = x.astype(np.float32)
y = y.astype(np.float32)

ceil = Ceil()
flat = tf.keras.layers.Flatten()
dense = tf.keras.layers.Dense(10)
softmax = tf.keras.layers.Softmax(-1)
inp = tf.keras.layers.Input((28, 28, 1))
conv = tf.keras.layers.Conv2D(1, 3, 1, 'same')

fx = conv(inp)
fx = ceil(fx)
fx = flat(fx)
fx = dense(fx)
fx = softmax(fx)
model = tf.keras.Model(inp, fx)
model.compile('adam', 'categorical_crossentropy')
model.fit(x, y, 100, 10)

**Comment by chunduriv** [comment]
@jvishnuvardhan, the issue was able to reproduce the issue on Colab using TF 2.7 , and TF-nightly(2.8.0-dev20211208). Please find the [gist here](https://colab.research.google.com/gist/chunduriv/2c86e6c6f84ebf0737a298ff2c7c93df/15716.ipynb#scrollTo=W7BruSKTPwFR) for reference.Thanks!

**Comment by jvishnuvardhan** [comment]
Triage notes: Here is the output

```
2.8.0-dev20211208
Epoch 1/10
WARNING:tensorflow:Gradients do not exist for variables ['conv2d/kernel:0', 'conv2d/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
WARNING:tensorflow:Gradients do not exist for variables ['conv2d/kernel:0', 'conv2d/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss`argument?
600/600 [==============================] - 9s 14ms/step - loss: 10.3094
Epoch 2/10
600/600 [==============================] - 8s 14ms/step - loss: 4.0716
```

**Comment by haifeng-jin** [comment]
Waiting for triage.
No gradients before `tf.math.ceil()`.

**Comment by haifeng-jin** [comment]
The currently implementation of the `tf.math.ceil()` is correct.
We don't expect the gradients to go over it.

If you have any special use case for it, you may use the following API to define a custom gradient.
https://www.tensorflow.org/api_docs/python/tf/custom_gradient
